contracts/actions/flashloan/FLAction.sol

```diff
@@ -0,0 +1,314 @@
+import "../../interfaces/flashloan/IERC3156FlashBorrower.sol";
+import "../../interfaces/aaveV2/ILendingPoolV2.sol";
+import "../../interfaces/balancer/IFlashLoans.sol";
+import "../../interfaces/euler/IEulerMarkets.sol";
+import "../../interfaces/euler/IDToken.sol";
+
+import "../../core/strategy/StrategyModel.sol";
+
+import "./helpers/FLHelper.sol";
+
+/// @title Action that gets and receives FL from different variety of sources
+contract FLAction is ActionBase, ReentrancyGuard, IFlashLoanBase, StrategyModel, FLHelper, DSMath {
+    using TokenUtils for address;
+    using SafeMath for uint256;
+
+    /// @dev FL Initiator must be this contract
+    error UntrustedInitiator();
+    /// @dev Caller in these functions must be relevant FL source address
+    error UntrustedLender();
+
+    error WrongPaybackAmountError(); // Wrong FL payback amount sent
+
+    enum FLSource {
+        EMPTY,
+        AAVE,
+        BALANCER,
+        EULER,
+        MAKER
+    }
+
+    /// @dev Function sig of RecipeExecutor._executeActionsFromFL()
+    bytes4 public constant CALLBACK_SELECTOR =
+        bytes4(
+            keccak256(
+                "_executeActionsFromFL((string,bytes[],bytes32[],bytes4[],uint8[][]),bytes32)"
+            )
+        );
+    bytes4 public constant RECIPE_EXECUTOR_ID = bytes4(keccak256("RecipeExecutor"));
+
+    /// @inheritdoc ActionBase
+    function actionType() public pure override returns (uint8) {
+        return uint8(ActionType.FL_ACTION);
+    }
+
+    // solhint-disable-next-line no-empty-blocks
+    function executeActionDirect(bytes memory _callData) public payable override {}
+
+    /// @inheritdoc ActionBase
+    /// @notice This action doesn't use flParamGetterAddr and flParamGetterData
+    /// @notice flParamGetterData is used to choose between FL providers
+    function executeAction(
+        bytes memory _callData,
+        bytes32[] memory,
+        uint8[] memory,
+        bytes32[] memory
+    ) public payable override returns (bytes32) {
+        FlashLoanParams memory params = abi.decode(_callData, (FlashLoanParams));
+        FLSource flSource = FLSource(uint8(bytes1(params.flParamGetterData)));
+        handleFlashloan(params, flSource);
+
+        return bytes32(params.amounts[0]);
+    }
+
+    function handleFlashloan(FlashLoanParams memory _flParams, FLSource _source) internal {
+        if (_source == FLSource.AAVE) {
+            _flAave(_flParams);
+        } else if (_source == FLSource.BALANCER) {
+            _flBalancer(_flParams);
+        } else if (_source == FLSource.EULER) {
+            _flEuler(_flParams);
+        } else if (_source == FLSource.MAKER) {
+            _flMaker(_flParams);
+        }
+    }
+
+    /// @notice Gets a Fl from AaveV2 and returns back the execution to the action address
+    /// @param _flParams All the amounts/tokens and related aave fl data
+    function _flAave(FlashLoanParams memory _flParams) internal {
```

Comment:
If we want to future proof this (Aave V3 coming soon on mainnet?), we could send lending pool address as flParamGetterAddr in params, so this one function covers both Aave V2 and V3 on mainnet on release, but for now this is okay, as it will support both V2 on mainnet and v3 on L2 depending on which Lending pool address is in FLHelper

Reply:
This is ok for now